Updated yrs to 0.18.2

fbarbat opened the pull request:
This includes introducing a Subscription abstraction to get rid of the subscription IDs previously used that are not available anymore in Yrs.

https://github.com/y-crdt/yswift/issues/42

fbarbat commented on Package.swift:
```diff
@@ -47,13 +47,13 @@ if ProcessInfo.processInfo.environment["YSWIFT_LOCAL"] != nil {
-            checksum: "d2633bdb1e9f257cd56a852f360f0d0f4bc1615a4c34a05e76a2da2c430a0f98"
-    )
-}
+//} else {
```
Left this like this intentionally to not forget to update the xcframework before merging

heckj replied:
You should be able to go ahead and drop this back into place, as the CI system builds with the local build all the time, to allow us to test and validate changes - and we'll want likely want to separate all these changes from the process of cutting a new release, which updates the checksums, tags, etc for those various bits.

fbarbat commented on Sources/YSwift/YSubscription.swift:
```diff
@@ -0,0 +1,21 @@
+        self.subscription = subscription
+    }
+
+    public func cancel() {
```
Felt that leaving a cancel method publicly available was useful when explicitly canceling, to avoid warnings in Swift when the variable was assigned but not read.

fbarbat commented on Tests/YSwiftTests/YSubscriptionTests.swift:
```diff
@@ -0,0 +1,92 @@
+    
+    // This causes memory to keep growing and I am not sure why
+    // I checked that UniFFI YSubscription struct is being dropped accordingly
+    // I also verified that the memory increase also happened when we used subscriptionIds and unobserve
```
As I describe here, I am not sure why setting up observations increase memory, even before this change. Any thoughts on this?

Horusiath replied:
In new model subscriptions themselves keep strong reference to a callback, while internal callback manager keeps a list of weak references. These references are not released right away, but in one of two cases:

1. When we drop an iterator used for triggering the callbacks.
2. When we add a new callback: since it's happening when using atomic update, we need to prepare a copy of older callbacks, so we also get rid of the weak refs without the strong ones.

You can add a simple text.insert operation to confirm if callback trigger is performing the cleanup. You could also try to cancel subscription explicitly. If none of these works, then probably we have memory leak problem in core lib.

fbarbat replied:
I tried this in a test and it doesn't prevent memory going up.
```swift
    func test_allocateTooMany() {
        for i in 1...1000000000 {
            let s = text.observe { _ in }
            text.unobserve(s)
            
            if i % 1000000 == 0 {
                print("\(Date()) Iterations: \(i) Resident Size: \(residentSize() ?? 0)")
                text.insert("a", at: 0)
            }
        }
    }
```
Not sure if the leak is in Yswift side or Yrs though.

fbarbat replied:
Note the above test was run before this PR. For this PR, the equivalent test is this one:
```swift
    func testallocateTooMany() {
        for i in 1...1000000000 {
            let s = text.observe { _ in }
            s.cancel()
            
            if i % 1000000 == 0 {
                print("\(Date()) Iterations: \(i) Resident Size: \(residentSize() ?? 0)")
                text.insert("a", at: 0)
            }
        }
    }
    
```
In both memory keeps going up.

heckj replied:
I suspect you've spotted an issue outside of what changed in this PR - if you're game, let's separate the two and hunt the memory issue we're seeing here separately from these changes.

fbarbat commented on Tests/YSwiftTests/YUndoManagerTests.swift:
```diff
@@ -86,6 +88,9 @@ class YUndoManagerTests: XCTestCase {
         XCTAssert (try manager.redo())
         XCTAssertEqual(received.value, "A")
+        
+        on_added.cancel()
```
This was just to remove the warnings and leave warnings in zero ✨

fbarbat commented on lib/Cargo.toml:
```diff
@@ -16,7 +16,7 @@ path = "uniffi-bindgen.rs"
 
 [dependencies]
-yrs = "0.17.4"
+yrs = "0.18.2"
```
The rest of the PR is just trying to fix things that break after this version change

fbarbat commented on lib/src/array.rs:
```diff
@@ -184,23 +186,18 @@ impl YrsArray {
 
-    pub(crate) fn observe(&self, delegate: Box<dyn YrsArrayObservationDelegate>) -> u32 {
-        let subscription_id = self
+    pub(crate) fn observe(&self, delegate: Box<dyn YrsArrayObservationDelegate>) -> Arc<YSubscription> {
```
I am not sure about this Arc<YSubscription> return type. Again, I am not a Rust developer so I am not sure about the UniFFI generated code would not compile because it expected a type like this.

fbarbat replied:
@Horusiath what do you think about this? Do you think this is safe?

fbarbat commented on lib/src/subscription.rs:
```diff
@@ -0,0 +1,22 @@
+  }
+}
+
+// impl Drop for YSubscription {
```
I can delete this. This is when I was testing that this was being properly deinitialized.

heckj replied:
yeah, go ahead and drop it if you're not going to be using it down the road.

fbarbat commented on lib/src/undo.rs:
```diff
@@ -59,43 +60,28 @@ impl YrsUndoManager {
         let m = self.acquire_lock();
         let subscription = m.observe_item_added(move |_, e| {
-            e.item.meta = delegate.call(YrsUndoEvent::new(e), e.item.meta);
+            *e.meta_mut() = delegate.call(YrsUndoEvent::new(e), *e.meta_mut());
```
Is this ok? I don't really understand what's going on here, I just wanted it to compile and this looked reasonable but I have no evidence that this is working as expected (other than tests passing).

Horusiath replied:
Yes, it's fine. `e.meta_mut()` returns a mutable reference to meta object of event (which here IIRC is a memory address to an object). `*e.meta_mut() = ...` is overriding this meta reference with new value (basically reassignment).

This way you can attach meta objects to undo events and pop them later on ie. remember cursor position when `undo` is being performed and then restore that cursor position from meta object during `redo` action.

heckj reviewed (changes requested):
Thanks for the PR!

I'll greedily defer the Rust acceptability questions to @Horusiath , as feel very imposter every time I tackle something like that.

The two files in /scaffolding can be dumped from the PR, as they'll get regenerated by CI, but that doesn't strictly matter all that much - when we cut a release and punch in a new hash and XCFramework file during the release, they'll be regenerated as well.

I'll download the branch/PR and do some digging myself as to the memory leak. Off the top of my head, I suspect it's likely that because you made a YSubscription thing that passes over the Rust/Swift boundary, that this process is likely incurring a heavier cost, and I suspect related, but I'd need to dig a bit on the Swift side to see what I could find.

fbarbat commented:
Thank you @heckj! Reverted the files you mentioned.
Regariding the leak, it's worth noting that the leak was already there before this PR. Specifically, I ran this in current master and I would see the same behavior. So I am not sure my changes introduced the leak. As shown in some Rust print lines over there, I made sure my subscriptions were dropped. If I was doing something wrong please let me know! Thanks

```swift
//
//  YSubscriptionTests.swift
//
//
//  Created by Fernando Barbat on 28/3/24.
//

import Foundation

import Foundation

import Combine
import XCTest
@testable import YSwift

class YSubscriptionTests: XCTestCase {
    var document: YDocument!
    var text: YText!
    
    override func setUp() {
        document = YDocument()
        text = document.getOrCreateText(named: "test")
    }
    
    override func tearDown() {
        document = nil
        text = nil
    }
    
    
    // This causes memory to keep growing and I am not sure why
    // I checked that UniFFI YSubscription struct is being dropped accordingly
    func test_allocateTooMany() {
        for i in 1...1000000000 {
            let s = text.observe { _ in }
            text.unobserve(s)
            
            if i % 1000000 == 0 {
                print("\(Date()) Iterations: \(i) Resident Size: \(residentSize() ?? 0)")
            }
        }
        
        print("Done!")
    }
    
    private func residentSize() -> UInt64? {
        var info = mach_task_basic_info()
        let MACH_TASK_BASIC_INFO_COUNT = MemoryLayout<mach_task_basic_info>.stride/MemoryLayout<natural_t>.stride
        var count = mach_msg_type_number_t(MACH_TASK_BASIC_INFO_COUNT)

        let kerr: kern_return_t = withUnsafeMutablePointer(to: &info) {
            $0.withMemoryRebound(to: integer_t.self, capacity: MACH_TASK_BASIC_INFO_COUNT) {
                task_info(mach_task_self_,
                          task_flavor_t(MACH_TASK_BASIC_INFO),
                          $0,
                          &count)
            }
        }

        guard kerr == KERN_SUCCESS else {
            return nil
        }
        
        return info.resident_size
    }
}
```

fbarbat commented:
@heckj exciting that this was approved! CI is failing though, how can we fix that?
Also, should we wait for @Horusiath comment about Rust Arc? All other conversations were resolved.
```
Working copy of https://github.com/y-crdt/yswift resolved at 0.2.0
error: local binary target 'yniffiFFI' at '/Users/runner/work/yswift/yswift/examples/Tuist/Dependencies/SwiftPackageManager/.build/checkouts/yswift/lib/yniffiFFI.xcframework' does not contain a binary artifact.
error: fatalError
The 'swift' command exited with error code 1
Consider creating an issue using the following link: https://github.com/tuist/tuist/issues/new/choose
Error: Process completed with exit code 1.
```

heckj commented:
Back to fixing Tuist again? Damnit. I think it might be time to rip that out if it's not going to be stable. I'll work on getting a fix enabled there. And yeah, I'd still like @Horusiath 's input on that `Arc` question - I think it's OK (since it compiles, which is a kind of crap answer), but I just don't know

heckj commented:
I couldn't update your branch, so I've opened https://github.com/y-crdt/yswift/pull/46, which will hopefully flow though smoothly. If that works, we can rebase this over latest main and see how it goes. I'm not 100% that this solves things, but I'm hopeful.

heckj commented:
@fbarbat Rebase this on the main branch, and it should be happier now... and I probably screwed up the current main branch, so if you need or want to regenerate the scaffold files to sort the rebasing, go for it - we'll get it fully sorted when we cut a release with this included.

fbarbat commented:
Rebased, kept your changes and updated the PR! Wouldn't it be nice to remove the generated files from GitHub, add them to git ignore, and make Xcode to generate them on build?

heckj commented:
Normally, I'd love to, but the content in the scaffold files is tightly coupled to the XCFramework binary that's generated for the same, so for someone "coming along" and just wanting to use the code, they need to be there. For releases, we update everything in lockstep. For latest development on the `main` branch, I'm not as chuffed if it's "broken" and requires a developer to use the YSWIFT_LOCAL along with building their own XCFramework

heckj commented:
All looks good with CI, tests passing, etc - so I'll hold on merging until you give me a thumbs up, in case you want to ask further questions. Once we merge this, I'll go ahead and do the leg-work to cut an updated release for more general consumption.

fbarbat commented:
I just realized I pushed some generated files changes (bad merge probably). Should I remove those? You mentioned those are regenerated on release... What do you think?
On my side, it would be nice to have @Horusiath feedback on Rust Arc, otherwise all good. We've been running for 5 days this change on prod from a custom fork I created and everything is looking good.

fbarbat commented:
Ok, I removed the changes to keep the PR cleaner. Can you run CI again? Thanks!
